csharp/src/Ice/OutputStream.cs

```diff
@@ -1014,6 +1020,60 @@ public void WriteTaggedSerializable(int tag, object? o)
         // Other methods
         //
 
+        /// <summary>Writes a sequence of bits to the stream, and returns this sequence backed by the stream's buffer.
+        /// </summary>
+        /// <param name="bitLength">The minimum number of bits in the sequence.</param>
+        /// <returns>The bit sequence, with all bits set. The actual length of the sequence is a multiple of 8.
+        /// </returns>
+        public BitSequence WriteBitSequence(int bitLength)
+        {
+            Debug.Assert(bitLength > 0);
+            int size = (bitLength >> 3) + ((bitLength & 0x07) != 0 ? 1 : 0);
+
+            Position startPos = _tail;
+            bool patchStartPos = (startPos.Offset == _currentSegment.Count);
+
+            int remaining = _capacity - Size;
+            if (size > remaining)
+            {
+                int segmentSize = Math.Max(DefaultSegmentSize, _currentSegment.Count * 2);
+                segmentSize = Math.Max(size - remaining, segmentSize);
+                byte[] buffer = new byte[segmentSize];
+                _segmentList.Add(buffer);
+                _currentSegment = buffer;
+                _tail.Segment++;
+                _tail.Offset = size - remaining;
+                Size += size;
+            }
+            else
+            {
+                _tail.Offset += size;
+                Size += size;
+            }
+
+            if (patchStartPos)
+            {
+                // Move startPos to the first byte of the next segment
+                startPos.Segment++;
+                startPos.Offset = 0;
+            }
+
+            if (startPos.Segment == _tail.Segment)
+            {
+                var span = _currentSegment.AsSpan(startPos.Offset, _tail.Offset - startPos.Offset);
+                span.Fill(255);
+                return new BitSequence(span);
+            }
+            else
+            {
+                var firstSpan = _segmentList[startPos.Segment].AsSpan(startPos.Offset);
+                var secondSpan = _currentSegment.AsSpan(0, _tail.Offset);
+                firstSpan.Fill(255);
+                secondSpan.Fill(255);
+                return new BitSequence(firstSpan, secondSpan);
+            }
```

Comment:
Maybe we can simplify this a bit, see also the fix for `Expand`
```suggestion
            Expand(size);

            int remainig = _currentSegment.Count - _tail.Offset;
            if (size <= remainig)
            {
                Span<byte> span = _currentSegment.AsSpan(_tail.Offset, size);
                span.Fill(255);
                _tail.Offset += size;
                return new BitSequence(span);
            }
            else
            {
                Span<byte> firstSpan = _currentSegment.AsSpan(_tail.Offset);
                firstSpan.Fill(255);
                _currentSegment = _segmentList[^1];
                Span<byte> secondSpan = _currentSegment.AsSpan(0, size - remainig);
                secondSpan.Fill(255);
                _tail.Offset = size - remainig;
                return new BitSequence(firstSpan, secondSpan);
            }
```

Reply:
Your code is missing the `bool patchStartPos = (startPos.Offset == _currentSegment.Count);` logic as far as I can tell, i.e. we can end up with an empty (illegal?) firstSpan.

Reply:
The idea is that Expand move _tail to the first segment when the current has no remaining bytes

Reply:
I forget to actually update the _tail in Expand, should work with last fix for Expand

Reply:
Is your code missing a `Size += size`?

Reply:
Yes

Reply:
Fixed as suggested. Also added Asserts and updated `WriteByte` to take advantage of this enhanced `Expand`.